include/deal.II/differentiation/sd/symengine_scalar_operations.h

```diff
@@ -146,8 +146,317 @@ namespace Differentiation
+     *         SymmetricTensor<2,dim,Expression>(...));
+     * @endcode
+     *
+     * @tparam SymbolicType Any symbolic type that is understood by the
+     *         add_to_symbol_map() functions. This includes individual
+     *         Expression, std::vector<Expression>, as well as
+     *         Tensors and SymmetricTensors of Expressions.
+     */
+    template <typename SymbolicType, typename... Args>
+    types::substitution_map
+    make_symbol_map(const SymbolicType &symbol, const Args &... other_symbols);
+
+    /**
+     * A convenience function for an adding empty entry, with the key value
+     * given by @p symbol, to the symbolic map @p substitution_map.
+     *
+     * This function is guaranteed to create an ordering that is identical
+     * to the typical add_to_substitution_map() call that is used when
+     * constructing a map to to perform symbol substitution.
+     * It exists primarily to create an initial map that can be used in the
+     * optimize() call to a BatchOptimizer, specifically if the values that
+     * are to be substituted into the map are not known at the time that the
+     * symbols used to construct symbolic expressions are defined.
+     * This helps one conform to the requirement that the arguments sent into
+     * lambda and LLVM JIT compiled functions (created by optimizing symbolic
+     * expressions) (i) be the same, and (ii) have a constant ordering.
+     *
+     * @tparam ignore_invalid_symbols A template parameter that enforces whether
+     *         or not the @p symbol has to be a valid one or not. In the
+     *         overwhelming majority of cases, the default value of
+     *         <tt>false</tt> should be selected, with the result that an
+     *         exception will be thrown if the input @p symbolic is, in fact,
+     *         not a symbolic value or expression.
+     *         An exceptional case is, for example, when performing symbolic
+     *         assembly on a finite element level. When extracting the symbolic
+     *         equivalent of the shape function gradients using FEExtractors,
+     *         the returned tensor will have some <em>a priori</em> determined
+     *         zero-valued components. These trivial components are not valid
+     *         symbols (as they are not symbolic expressions), and we would
+     *         typically wish to guard against their (erroneous) inclusion. In
+     *         this scenario, for convenience, one could set
+     *         @p ignore_invalid_symbols to <tt>true</tt> and these zero-valued
+     *         entries would be skipped over and ignored.
+     *
+     * @note In this function, the @p ValueType is somewhat arbitrary as
+     * it is only used to create dummy values.
+     */
+    template <bool ignore_invalid_symbols = false, typename ValueType = double>
+    void
+    add_to_symbol_map(types::substitution_map &symbol_map,
+                      const Expression &       symbol);
```

Comment:
same question about symbol order here

Reply:
This is the case that dictates the necessary order of the arguments:
```c++
    template <bool ignore_invalid_symbols = false,
              typename SymbolicType,
              typename... Args>
    void
    add_to_symbol_map(types::substitution_map &symbol_map,
                      const SymbolicType &     symbol,
                      const Args &... other_symbols);
``` 
The parameter pack must be the last argument to the function, so its impossible to consistently have `symbol_map` at the last argument if you want to use this template magic and have all the functions feel like take arguments in the same order. So, to me, it makes no sense to interleave the input and output arguments.

Reply:
I see -- fair point!